src/math/log.c

```diff
@@ -0,0 +1,115 @@
+//
+//
+
+/* Include Files */
+#include <sof/math/log.h>
+
+/**
+ *  Base-2 logarithm
+ *
+ * Y = (u) computes the base-2 logarithm of
+ * u using lookup table.
+ * input u must be scalar/real number and positive
+ *
+ * +------------------+-----------------+--------+--------+
+ * | u		      | y (returntype)	|   u	 |  y	  |
+ * +----+-----+-------+----+----+-------+--------+--------+
+ * |WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ * +----+-----+-------+----+----+-------+--------+--------+
+ * | 32 | 1   |	 0    | 32 | 27 |   0	| 31.1	 | 5.27	  |
+ * +------------------+-----------------+--------+--------+
+ *
+ * Arguments	: uint32_t u
+ * u input range [1 to 2147483647.5]
+ *
+ * Return Type	: int32_t  y
+ * y output range [0 to 31]
+ *
+ */
+int32_t base2_logarithm(uint32_t u)
+{
+	static const int32_t iv1[129] = {
+	    0,	   736,	  1466,	 2190,	2909,  3623,  4331,  5034,  5732,  6425,  7112,	 7795,
+	    8473,  9146,  9814,	 10477, 11136, 11791, 12440, 13086, 13727, 14363, 14996, 15624,
+	    16248, 16868, 17484, 18096, 18704, 19308, 19909, 20505, 21098, 21687, 22272, 22854,
+	    23433, 24007, 24579, 25146, 25711, 26272, 26830, 27384, 27936, 28484, 29029, 29571,
+	    30109, 30645, 31178, 31707, 32234, 32758, 33279, 33797, 34312, 34825, 35334, 35841,
+	    36346, 36847, 37346, 37842, 38336, 38827, 39316, 39802, 40286, 40767, 41246, 41722,
+	    42196, 42667, 43137, 43603, 44068, 44530, 44990, 45448, 45904, 46357, 46809, 47258,
+	    47705, 48150, 48593, 49034, 49472, 49909, 50344, 50776, 51207, 51636, 52063, 52488,
+	    52911, 53332, 53751, 54169, 54584, 54998, 55410, 55820, 56229, 56635, 57040, 57443,
+	    57845, 58245, 58643, 59039, 59434, 59827, 60219, 60609, 60997, 61384, 61769, 62152,
+	    62534, 62915, 63294, 63671, 64047, 64421, 64794, 65166, 65536};
+	static const int8_t iv[256] = {
+	    8, 7, 6, 6, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+	    3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+	    2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+	    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+	    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+	uint64_t v;
+	uint32_t x;
+	int num_left_shifts;
+	int shift_factor;
+	unsigned int slice_temp;
+	/*  Skipping for-loop */
+	/*  Unroll the loop so there will be no branching. */
+	/*  Loop unroll execute 20% faster (CPU cycle count )than for loop */
+	/*  Every iteration ,look for leading zeros are in the high */
```

Comment:
typo after "iteration" should be comma then space, and by high do you mean most significant?

Reply:
I will correct the typo . 

_by high do you mean most significant_
Yes. Get shift bits from the lookup table . Iterate over `u`  input upper nibble bits [ keep doing left shift ]  . Book shelve left shift  score so the most significant bits is a 1.